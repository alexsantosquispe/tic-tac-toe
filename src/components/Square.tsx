import clsx from 'clsx';
import { memo, type ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';
import { CircleIcon } from '../icons/CircleIcon';
import { XIcon } from '../icons/XIcon';
import type { SquareValueTypes } from '../types';

interface SquareProps {
  value: SquareValueTypes;
  onClickItem: () => void;
  highLight?: boolean;
  isDisabled?: boolean;
}

const SQUARE_ICONS: Record<'X' | 'O', ReactNode> = {
  X: <XIcon className="size-20 text-rose-600 md:size-28" />,
  O: <CircleIcon className="size-16 text-slate-600 md:size-20" />
};

const Square = ({
  value,
  onClickItem,
  highLight = false,
  isDisabled = false
}: SquareProps) => {
  const icon = value ? SQUARE_ICONS[value] : null;
  const disabled = !!value || isDisabled;

  return (
    <button
      onClick={onClickItem}
      disabled={disabled}
      className={twMerge(
        'flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150',
        clsx({
          'hover:cursor-pointer hover:bg-neutral-50': !disabled,
          'hover:cursor-not-allowed': disabled,
          'bg-rose-50': highLight
        })
      )}
      aria-disabled={disabled}
      aria-label={value || 'Empty'}
      type="button"
    >
      {icon}
    </button>
  );
};

export default memo(Square);
