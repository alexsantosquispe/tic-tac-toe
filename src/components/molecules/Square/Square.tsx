import { AnimatePresence, motion } from 'framer-motion';
import { memo, useCallback } from 'react';

import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';
import { CircleIcon } from '../../../icons/CircleIcon';
import { XIcon } from '../../../icons/XIcon';
import type { SquareValueTypes } from '../../../types';

const ICONS = {
  X: {
    component: XIcon,
    className: 'size-28 md:size-36'
  },
  O: {
    component: CircleIcon,
    className: 'size-20 md:size-28'
  }
} as const;

const COLORS = {
  X: 'text-rose-600',
  O: 'text-neutral-600 dark:text-neutral-400',
  hover: 'text-neutral-200/60 dark:text-white/20'
} as const;

const renderIcon = (player: 'X' | 'O', color: string) => {
  const { component: Icon, className } = ICONS[player];
  return <Icon className={twMerge(className, color)} />;
};

interface SquareProps {
  index: number;
  value: SquareValueTypes;
  onClickItem: (index: number) => void;
  currentPlayer: 'X' | 'O';
  highLight?: boolean;
  isDisabled?: boolean;
}

const Square = ({
  index,
  value,
  onClickItem,
  currentPlayer,
  highLight = false,
  isDisabled = false
}: SquareProps) => {
  const disabled = !!value || isDisabled;
  const icon = value !== '' ? renderIcon(value, COLORS[value]) : null;
  const hoverIcon =
    !icon && !disabled ? renderIcon(currentPlayer, COLORS.hover) : null;

  const handleClick = useCallback(() => {
    onClickItem(index);
  }, [index, onClickItem]);

  return (
    <button
      data-testid={`square-${index}`}
      onClick={handleClick}
      disabled={disabled}
      className={twMerge(
        'group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20',
        clsx({
          'hover:cursor-pointer hover:bg-neutral-50 dark:hover:bg-white/10':
            !disabled,
          'hover:cursor-not-allowed': disabled,
          'bg-rose-50 dark:bg-rose-900/30': highLight
        })
      )}
      aria-label={value || 'Empty'}
      type="button"
    >
      <AnimatePresence>
        {icon && (
          <motion.span
            key={value}
            initial={{ scale: 0.5, opacity: 0 }}
            animate={{ scale: 1, opacity: 1 }}
            transition={{ duration: 0.15, ease: 'easeOut' }}
          >
            {icon}
          </motion.span>
        )}
      </AnimatePresence>

      {hoverIcon && (
        <span className="absolute inset-0 flex items-center justify-center opacity-0 transition-opacity duration-150 group-hover:opacity-100">
          {hoverIcon}
        </span>
      )}
    </button>
  );
};

export default memo(Square);
