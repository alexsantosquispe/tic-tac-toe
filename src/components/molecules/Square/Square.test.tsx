import '@testing-library/jest-dom';

import { fireEvent, render, screen, within } from '@testing-library/react';

import type { ComponentProps } from 'react';
import Square from './Square';

describe('Square', () => {
  const onClickItemMock = jest.fn();

  const props: ComponentProps<typeof Square> = {
    index: 0,
    value: '',
    onClickItem: onClickItemMock,
    currentPlayer: 'X',
    highLight: false,
    isDisabled: false
  };

  describe('styles', () => {
    it('should render the square with the correct styles', () => {
      let component = render(<Square {...props} />);
      expect(component).toMatchSnapshot();

      component = render(<Square {...props} value="O" />);
      expect(component).toMatchSnapshot();

      component = render(<Square {...props} value="X" />);
      expect(component).toMatchSnapshot();
    });

    it('should render a placeholder icon on hover when the square is empty', () => {
      render(<Square {...props} />);

      const square = screen.getByRole('button');

      fireEvent.mouseEnter(square);

      expect(square).toHaveClass('hover:bg-neutral-50');

      expect(within(square).getByTestId('x-icon')).toBeInTheDocument();
    });

    it('should display highlight styles if the square is a highlight winner combination', () => {
      render(
        <Square {...props} value="X" isDisabled={true} highLight={true} />
      );

      const squareButton = screen.getByRole('button');

      expect(squareButton).toBeDisabled();

      expect(squareButton).toHaveClass('hover:cursor-not-allowed');

      expect(squareButton).toHaveClass('bg-rose-50 dark:bg-rose-800/40');
    });

    it('should display disabled styles if the square is not a highlight winner combination', () => {
      render(
        <Square {...props} value="X" isDisabled={true} highLight={false} />
      );

      const squareButton = screen.getByRole('button');

      expect(squareButton).toBeDisabled();

      expect(squareButton).toHaveClass('hover:cursor-not-allowed');

      expect(squareButton).toHaveClass(
        'opacity-25 brightness-30 dark:opacity-80 dark:brightness-50'
      );
    });
  });

  describe('behavior', () => {
    it('should call onClickItem when the square is clicked', () => {
      render(<Square {...props} />);

      const squareButton = screen.getByRole('button');

      fireEvent.click(squareButton);

      expect(onClickItemMock).toHaveBeenCalledTimes(1);

      expect(onClickItemMock).toHaveBeenCalledWith(props.index);
    });

    it('should not call onClickItem when the square it was already selected', () => {
      render(<Square {...props} value="X" />);

      const squareButton = screen.getByRole('button');

      expect(squareButton).toBeDisabled();

      fireEvent.click(squareButton);

      expect(onClickItemMock).toHaveBeenCalledTimes(0);

      expect(squareButton).toHaveClass('hover:cursor-not-allowed');
    });
  });
});
