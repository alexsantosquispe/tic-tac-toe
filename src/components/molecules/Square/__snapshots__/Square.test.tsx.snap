// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Square styles should render the square with the correct styles 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-label="Empty"
        class="group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20 hover:cursor-pointer hover:bg-neutral-50 dark:hover:bg-white/10"
        data-testid="square-0"
        type="button"
      >
        <span
          class="absolute inset-0 flex items-center justify-center opacity-0 transition-opacity duration-150 group-hover:opacity-100"
        >
          <svg
            class="size-28 md:size-36 text-neutral-200/60 dark:text-white/20"
            data-testid="x-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-label="Empty"
      class="group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20 hover:cursor-pointer hover:bg-neutral-50 dark:hover:bg-white/10"
      data-testid="square-0"
      type="button"
    >
      <span
        class="absolute inset-0 flex items-center justify-center opacity-0 transition-opacity duration-150 group-hover:opacity-100"
      >
        <svg
          class="size-28 md:size-36 text-neutral-200/60 dark:text-white/20"
          data-testid="x-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="3"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Square styles should render the square with the correct styles 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-label="Empty"
        class="group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20 hover:cursor-pointer hover:bg-neutral-50 dark:hover:bg-white/10"
        data-testid="square-0"
        type="button"
      >
        <span
          class="absolute inset-0 flex items-center justify-center opacity-0 transition-opacity duration-150 group-hover:opacity-100"
        >
          <svg
            class="size-28 md:size-36 text-neutral-200/60 dark:text-white/20"
            data-testid="x-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </span>
      </button>
    </div>
    <div>
      <button
        aria-label="O"
        class="group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20 hover:cursor-not-allowed"
        data-testid="square-0"
        disabled=""
        type="button"
      >
        <span
          style="opacity: 0; transform: scale(0.5);"
        >
          <svg
            class="size-20 md:size-28 text-neutral-600 dark:text-neutral-200"
            data-testid="circle-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
          </svg>
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-label="O"
      class="group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20 hover:cursor-not-allowed"
      data-testid="square-0"
      disabled=""
      type="button"
    >
      <span
        style="opacity: 0; transform: scale(0.5);"
      >
        <svg
          class="size-20 md:size-28 text-neutral-600 dark:text-neutral-200"
          data-testid="circle-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="3"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
        </svg>
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Square styles should render the square with the correct styles 3`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-label="Empty"
        class="group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20 hover:cursor-pointer hover:bg-neutral-50 dark:hover:bg-white/10"
        data-testid="square-0"
        type="button"
      >
        <span
          class="absolute inset-0 flex items-center justify-center opacity-0 transition-opacity duration-150 group-hover:opacity-100"
        >
          <svg
            class="size-28 md:size-36 text-neutral-200/60 dark:text-white/20"
            data-testid="x-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </span>
      </button>
    </div>
    <div>
      <button
        aria-label="O"
        class="group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20 hover:cursor-not-allowed"
        data-testid="square-0"
        disabled=""
        type="button"
      >
        <span
          style="opacity: 0; transform: scale(0.5);"
        >
          <svg
            class="size-20 md:size-28 text-neutral-600 dark:text-neutral-200"
            data-testid="circle-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
          </svg>
        </span>
      </button>
    </div>
    <div>
      <button
        aria-label="X"
        class="group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20 hover:cursor-not-allowed"
        data-testid="square-0"
        disabled=""
        type="button"
      >
        <span
          style="opacity: 0; transform: scale(0.5);"
        >
          <svg
            class="size-28 md:size-36 text-rose-600 dark:text-rose-500"
            data-testid="x-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-label="X"
      class="group relative flex h-1/3 w-1/3 items-center justify-center border border-dashed border-neutral-200 p-1 transition-colors duration-150 dark:border-white/20 hover:cursor-not-allowed"
      data-testid="square-0"
      disabled=""
      type="button"
    >
      <span
        style="opacity: 0; transform: scale(0.5);"
      >
        <svg
          class="size-28 md:size-36 text-rose-600 dark:text-rose-500"
          data-testid="x-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="3"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
